File: core/src/main/java/dev/morphia/mapping/codec/kotlin/ReadWritePropertyCodec.java
package dev.morphia.mapping.codec.kotlin;

import kotlin.properties.ReadWriteProperty;
import org.bson.BsonReader;
import org.bson.BsonWriter;
import org.bson.codecs.Codec;
import org.bson.codecs.DecoderContext;
import org.bson.codecs.EncoderContext;

/**
 * @morphia.internal
 * @since 2.2
 */
@SuppressWarnings("unchecked")
public class ReadWritePropertyCodec implements Codec<Object> {
    private final Codec<Object> codec;

    public ReadWritePropertyCodec(Codec<?> codec) {
        this.codec = (Codec<Object>) codec;
    }

    @Override
    public Object decode(BsonReader reader, DecoderContext context) {
        return codec.decode(reader, context);
    }

    @Override
    public void encode(BsonWriter writer, Object value, EncoderContext context) {
        context.encodeWithChildContext(codec, writer, value);
    }

    @Override
    public Class<Object> getEncoderClass() {
        return (Class<Object>) ((Class) ReadWriteProperty.class);
    }
}

File: core/src/main/kotlin/dev/morphia/mapping/codec/kotlin/ReadWritePropertyCodec.kt
package dev.morphia.mapping.codec.kotlin

import org.bson.BsonReader
import org.bson.BsonWriter
import org.bson.codecs.Codec
import org.bson.codecs.DecoderContext
import org.bson.codecs.EncoderContext
import kotlin.properties.ReadWriteProperty
import kotlin.reflect.KProperty

@Suppress("UNCHECKED_CAST")
class ReadWritePropertyCodec<T>(private val codec: Codec<Any>, private val type: Class<T>) : Codec<Any> {
    private lateinit var property: KProperty<T>
    override fun getEncoderClass(): Class<Any> = ReadWriteProperty::class.java as Class<Any>
    override fun encode(writer: BsonWriter, value: Any, context: EncoderContext) {
        context.encodeWithChildContext(codec, writer, value)
    }

    override fun decode(reader: BsonReader, context: DecoderContext): Any? {
        return codec.decode(reader, context)
    }
}
